### PR TITLE
Remove ${FOLLY_WITH_DEPENDENCIES} and fix Boost::<component> links

### PR DESCRIPTION
**Pull request by assignUser**
This PR removes the variable `FOLLY_WITH_DEPENDENCIES` that was used to  link to folly instead of the more appropriate `Folly::folly` target.

A consequence of this change was the proper linking to Boost targets in the form of `Boost::<component>` instead of `BOOST_LIBRARIES` or `BOOST_<component>_LIBRARIES`, this in turn made some changes to the the folly and boost `resolve_dependency` modules necessary:
- folly and Boost are now built within a subdirectory. This creates a new, smaller scope. Previously they were built in the root scope. This allows us to use directory scoped thing like `add_compile_options` without affecting the main build.
- The FindBoost module and the BoostConfig.cmake create a `Boost::headers` target that contains the include dir with the header only components, there are no single targets for each header library (at least in some installations). Building Boost from source does not create this target but rather creates a target for each library. To align both versions of using Boost the new `FindBoost.cmake` will create the Boost::headers target so we can link to that uniformly if we want to use it.

This is to be merged after #4689 (this contains the commits from that PR which I will drop via rebase once it is merged)

### CODE REVIEW

**Review by assignUser — COMMENTED** [review]
These where messing with the ripgrep filters, which are supposed to match .gitignore 1:1. I tested it and git did not ignore the file (e.g. velox/common/config/Config.h) but that might be version dependent. 

So better to be save than sorry?

File: .gitignore
Comment (assignUser) [inline .gitignore:1]: These where messing with the ripgrep filters, which are supposed to match .gitignore 1:1. I tested it and git did not ignore the file (e.g. velox/common/config/Config.h) but that might be version dependent.

So better to be save than sorry?

File: CMake/resolve_dependency_modules/boost/FindBoost.cmake.in
Comment (assignUser) [inline CMake/resolve_dependency_modules/boost/FindBoost.cmake.in:1]: This will dynamically generate the FindBoost.cmake so we don't have to manually add the targets in case we add new components to `BOOST_INCLUDE_LIBRARY`

File: .gitignore [hunk 1]
```diff
@@ -86,6 +88,7 @@ _build/
 *.pdf
 *.swp
 a.out
+CMake/resolve_dependency_module/boost/FindBoost.cmake
```
Comment (assignUser) [inline .gitignore:91]: As this is now regenerated on each configure run we can ignore it.

File: docker-compose.yml [hunk 2]
```diff
@@ -50,3 +50,22 @@ services:
     volumes:
       - .:/velox:delegated
     command: /bin/bash -c "scl enable gcc-toolset-9 '/velox/scripts/docker-command.sh'"
+
+  python:
```
Comment (pedroerp) [inline docker-compose.yml:54]: this looks unrelated to this change?

**Review by pedroerp — APPROVED** [review]
Thanks @assignUser, the changes look good. Just have a small comment about the docker file change.

File: docker-compose.yml (hunk 2 above)
Comment (assignUser) [reply docker-compose.yml:54]: Ah, yes I used this for testing these changes and found it useful. We can move into a separate PR if you prefer?

Comment (pedroerp) [reply docker-compose.yml:54]: I think that's ok. Just making sure this isn't leftover and we still want this merged?

Comment (assignUser) [reply docker-compose.yml:54]: I will uses this in future testing so it will likely also be useful for others. :+1:

**Comment by kgpai** [comment]
@assignUser Can you validate that this works with presto and doesnt break them before we merge ?

**Comment by assignUser** [comment]
@kgpai It is still very unergonomic to test these changes... it configures fine but there seems to be an incompatibility while building with recent changes in velox (unrelated to this PR)?
```
ests/CMakeFiles/presto_operators_test.dir/PlanNodeSerdeTest.cpp.o -MF presto_cpp/main/operators/tests/CMakeFiles/presto_operators_test.dir/PlanNodeSerdeTest.cpp.o.d -o presto_cpp/main/operators/tests/CMakeFiles/presto_operators_test.dir/PlanNodeSerdeTest.cpp.o -c ../../presto_cpp/main/operators/tests/PlanNodeSerdeTest.cpp
../../presto_cpp/main/operators/tests/PlanNodeSerdeTest.cpp: In constructor ‘facebook::velox::exec::test::PlanNodeSerdeTest::PlanNodeSerdeTest()’:
../../presto_cpp/main/operators/tests/PlanNodeSerdeTest.cpp:106:18: error: ‘registerSerDe’ is not a member of ‘facebook::velox::exec::test::PlanBuilder’
  106 |     PlanBuilder::registerSerDe();
```

**Comment by kgpai** [comment]
Yes we need to define a strong interface so we can have some backwards compatibility across velox/presto. I suppose since it configures fine, it should probably build fine (once the c++ issues are resolved).

**Comment by laithsakka** [comment]
@assignUser  nightly fuzzer runs has been failing to build since May 5. this is the only PR that touch build files on that date?
its probably the reason for the block can you take a look?

File: CMake/resolve_dependency_modules/icu.cmake [hunk 3]
```diff
@@ -75,7 +75,7 @@ foreach(component ${icu_components})
   set_target_properties(
     ICU::${component}
     PROPERTIES IMPORTED_LOCATION ${ICU_${component}_LIBRARY}
-               INTERFACE_INCLUDE_DIRECTORIES ${ICU_INCLUDE_DIRS})
+               INTERFACE_SYSTEM_INCLUDE_DIRECTORIES ${ICU_INCLUDE_DIRS})
```
Comment (zhztheplayer) [inline CMake/resolve_dependency_modules/icu.cmake:78]: @assignUser Do you remember the reason of this change?

This makes `ICU::uc` not propagate the headers to dependent targets like `re2`.

Comment (assignUser) [reply CMake/resolve_dependency_modules/icu.cmake:78]: > Projects may also get and set the property directly, but must be aware that adding directories to this property does not make those directories used during compilation.

[docs](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_SYSTEM_INCLUDE_DIRECTORIES.html)

Hm seems we have to set both properties? This has been in there a long time, where are you seeing the error?

Comment (assignUser) [reply CMake/resolve_dependency_modules/icu.cmake:78]: the reason to use it is to avoid warnings in icu code tripping -Werror and blowing up a build
